### DISCUSSION TITLE
Truststore configuration (contradictionary options in guides)

### DISCUSSION THREAD

**Discussion by dasniko** [keycloak/keycloak#12310] [Archive]
Hi,
In the Keycloak Guides, there are two different options mentioned for configuring the truststore, which is used on outgoing https requests:

1. [Configuring TLS / Using a truststore](https://www.keycloak.org/server/enabletls#_using_a_truststore)
This is using the `https-trust-store...` config, here is mentioned _"This truststore is used for outgoing https requests as well as mTLS requests."_
2. [Configuring outgoing HTTP requests](https://www.keycloak.org/server/outgoinghttp#_outgoing_https_request_truststore)
This guide mentions the `spi-truststore-file-...` config for outgoing https requests.

Which is correct?
It seems that only the latter is working properly with an `ldaps://` connection.
In our project, we have an extension using the `SimpleHttp` class, but outgoing https requests fail with problems using the cert path. The logs say that no proper cert can be found, but it is actually in the truststore we've configured with the `spi-truststore-file...` aproach.

We can't use the `https-trust-store...` config approach, as it fails to start Keycloak when having a custom keystore configured at the same time (with `https-key-store-file`).

Is anybody able to explain for which approach I do have to use which config approach?
And does the `SimpleHttp` class really use the truststore configured via the SPI config approach?
I'm a bit lost and don't know where to look else to get it work.

**Comment by dasniko** [comment]
Ok, https://github.com/keycloak/keycloak/pull/12261 makes it explit, that the `spi-truststore-file-...` approach is the correct one.

But still, why is there also `https-trust-store-...` config and when should/can I use it?

**Reply by pedroigor** [reply]
The difference is where they are used. And I think you have a point and we can possibly improve this.

The HTTPS truststore is for enabling TLS to Vert.X, a server-truststore for trusting client certificates and certificate chains. And the Keycloak Truststore is a client trust store, where the server is acting as a client and establishing TLS connections with external services such as LDAP, SMTP, or HTTPS.

I think this separation still makes sense because they are targeted to different areas. But perhaps we can default to using the HTTPS truststore as a Keycloak Truststore if the latter is not explicitly set. Perhaps this makes things even more confusing too, because they are indeed distinct truststores.

**Comment by dasniko** [comment]
Thanks @pedroigor for explaining that. It wasn't clear to me, that there are two different truststores being used for different usecases. This might be a bit confusing and misleading. Perhaps this should be also made more explicit and pointed out in the docs. A cross-reference from the "Enabling TLS" guide to the new "Configuring the truststore" guide might be a good idea.

I do not necessarily need a fallback solution, or "use the former if the latter is not explictly set", but a clear statement and explanation in the docs would be sufficient (at least for me, don't know how others feel/think about it).

**Comment by jeffrson** [comment]
Unfortunately, neither https://www.keycloak.org/server/all-config nor `kc.sh start -help` mentions 'spi-truststore-file-...'. Also the server configuration guide has "Configuring TLS" vs "Configuring a Truststore" where it is not quite obvious to choose the latter (even if the smaller description would hint).

I'm glad this was mentioned here (thank you @dasniko) - I was desperately trying to import my CA into 'https-trust-store...' or even JVMs truststore to no avail.

**Reply by pedroigor** [reply]
Looks like we should rename the `Configuring a Truststore` guide to something more meaningful such as `Configuring the server to communicate over TLS`. Or something better that clearly indicates the purpose of the truststore configuration.

@stianst @vmuzikar @jeffrson @dasniko wdyt?

**Reply by vmuzikar** [reply]
Let me answer with a question. 🙂 Do we really need two truststores? Are there any use-cases for it? I propose to unify the truststores configuration (even though it'd still be technically two truststores underneath). Alternatively, keep the two as they are right now but introduce a new config option that would configure both in a unified manner.

See also related: https://github.com/keycloak/keycloak/issues/9784

**Reply by pedroigor** [reply]
Configuration-wise, we could use a single option. However, I'm not sure how that would impact people potentially forced to have separate trust stores as they hold certificates for different purposes. IIRC, there is also a limitation on JDK when dealing with a huge number of certs in a truststore as per https://issues.redhat.com/browse/WFLY-11007.

I'm not 100% sure how much a single option will make things easier because we still have to explain it. Perhaps improving the guides is enough for now, at least as a first try before changing our configuration.

**Reply by dasniko** [reply]
In my (very personal) opinion, there's no need for having two different truststores, at least at the logical/configuration level. None of my clients needs two separated truststores. And also with Wildfly, everything worked pretty well with only one truststore.

If there really will exist two truststores with a separate configuration, like it is currently the case, perhaps renaming the guides to something like "Using TLS certificates for INCOMING requests" and "...OUTGOING requests" would point the users more precicely to the proper guide!?

**Reply by pedroigor** [reply]
But both trust stores were there too in legacy.

Based on your suggestion, I think we should then rename the guide to `Using TLS for outgoing connections`?

**Reply by dasniko** [reply]
> But both trust stores were there too in legacy.

Oh, really? This is also new to me. Was it even documented?
So, apparently nobody used them and only used one truststore for storing the certs for outgoing connections and incoming X.509 certs... and everything worked. 🤷‍♂️

**Reply by pedroigor** [reply]
The documentation is here https://www.keycloak.org/docs/latest/server_installation/#_truststore.

Perhaps Wildfly has some black magic to make a truststore defined within the security subsystem available to applications. But from my experience, you usually need that SPI for outgoing TLS connections.

**Reply by dasniko** [reply]
> The documentation is here https://www.keycloak.org/docs/latest/server_installation/#_truststore.
> 
> 
> 
> Perhaps Wildfly has some black magic to make a truststore defined within the security subsystem available to applications. But from my experience, you usually need that SPI for outgoing TLS connections.

Ah, ok, now... seems that I've fallen into an error 🤦‍♂️ Bummer!

Like most companies I know of, they and I used the Keycloak Docker image, which has in the legacy Wildfly distribution this pretty awesome convenience script: https://github.com/keycloak/keycloak-containers/blob/main/server/tools/x509.sh
This script does the "magic" and configures the SPI for the outgoing AND the subsystem for the incoming requests with the same truststore, which is being built-up on demand when executing the script. This leads to the assumption, there is only one truststore...

If we could have a similar behavior (one truststore being used at both places) for the Quarkus distribution as a default, I would prefer that. Optionally it could be possible to configure two separated truststores, in case anyone needs that. This would be pretty awesome!

**Comment by pedroigor** [comment]
@dasniko @jeffrson Could you please take a look at https://github.com/keycloak/keycloak/pull/14961?

Even though a very minor change, for me it now looks more clear.

**Reply by dasniko** [reply]
See comment inline in the PR.

**Comment by psytester** [comment]
While setting up CRL check I catched this lack of documentation.
For CRL check the `spi-truststore-file` needs to be used instead of `https-trust-store-file`.
But CRL check is no "outgoing TLS connection"

The documentation is not clear at this point, but this was my last puzzle to overcome `truststore not available`.
